src/test/java/org/jenkinsci/plugins/workflow/steps/durable_task/ShellStepTest.java

```diff
@@ -459,6 +475,102 @@ private Object writeReplace() {
         }
     }
 
+    @Issue("JENKINS-54133")
+    @Test public void remoteConsoleNotes() throws Exception {
+        assumeFalse(Functions.isWindows()); // TODO create Windows equivalent
+        j.createSlave("remote", null, null);
+        WorkflowJob p = j.jenkins.createProject(WorkflowJob.class, "p");
+        p.setDefinition(new CpsFlowDefinition(
+            "node('master') {\n" +
+            "  markUp {\n" +
+            "    sh 'echo hello from master'\n" +
+            "  }\n" +
+            "}\n" +
+            "node('remote') {\n" +
+            "  markUp {\n" +
+            "    sh 'echo hello from agent'\n" +
+            "  }\n" +
+            "  markUp(smart: true) {\n" +
+            "    sh 'echo hello from halfway in between'\n" +
+            "  }\n" +
+            "}", true));
+        logging.recordPackage(ConsoleNote.class, Level.FINE);
+        WorkflowRun b = j.assertBuildStatusSuccess(p.scheduleBuild2(0));
+        b.getLogText().writeRawLogTo(0, System.err);
+        StringWriter w = new StringWriter();
+        b.getLogText().writeHtmlTo(0, w);
+        assertThat("a ConsoleNote created in the master is trusted", w.toString(), containsString("<b>hello</b> from master"));
+        assertThat("but this one was created in the agent and is discarded", w.toString(), containsString("hello from agent"));
+        assertThat("however we can pass it from the master to agent", w.toString(), containsString("<b>hello</b> from halfway in between"));
+    }
+    public static final class MarkUpStep extends Step {
+        @DataBoundSetter public boolean smart;
+        @DataBoundConstructor public MarkUpStep() {}
+        @Override public StepExecution start(StepContext context) throws Exception {
+            return new Exec(context, smart);
+        }
+        private static final class Exec extends StepExecution {
+            final boolean smart;
+            Exec(StepContext context, boolean smart) {
+                super(context);
+                this.smart = smart;
+            }
+            @Override public boolean start() throws Exception {
+                getContext().newBodyInvoker().
+                    withContext(BodyInvoker.mergeConsoleLogFilters(getContext().get(ConsoleLogFilter.class), new Filter(smart))).
+                    withCallback(BodyExecutionCallback.wrap(getContext())).
+                    start();
+                return false;
+            }
+        }
+        private static final class Filter extends ConsoleLogFilter implements Serializable {
+            private final @CheckForNull byte[] note;
+            Filter(boolean smart) throws IOException {
+                JenkinsJVM.checkJenkinsJVM();
+                if (smart) {
+                    ByteArrayOutputStream baos = new ByteArrayOutputStream();
+                    new HelloNote().encodeTo(baos);
+                    note = baos.toByteArray();
```

Comment:
The `smart` approach feels terribly hacky. If we are going to use it in actual code as in jenkinsci/ant-plugin#32, is there any documentation we can update (maybe `ConsoleLogFilter` or `ConsoleNote` itself) to clearly explain the problem with using `ConsoleNote` remotely, describe the pre-generation hack, and describe the preferred way to accomplish the same result for new code (Which I think is `ConsoleAnnotatorFactory` if I understand [your comment on JENKINS-54133](https://issues.jenkins-ci.org/browse/JENKINS-54133?focusedCommentId=351673&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-351673) correctly.)

Reply:
Documented in https://github.com/jenkinsci/jenkins/pull/3703/commits/8d73038a27420fdee0f5b4f97b47229c243345e8.

Reply:
(prepared in a core PR)

Reply:
https://github.com/jenkinsci/jenkins/pull/3703 for a link to the PR.